exec/java-exec/src/main/java/org/apache/drill/exec/planner/cost/DrillRelMdSelectivity.java

```diff
@@ -137,34 +138,33 @@ private double getScanSelectivityInternal(DrillTable table, RexNode predicate, L
     for (RexNode pred : RelOptUtil.conjunctions(predicate)) {
       double orSel = 0;
       for (RexNode orPred : RelOptUtil.disjunctions(pred)) {
-        //CALCITE guess
-        Double guess = RelMdUtil.guessSelectivity(pred);
+
         if (orPred.isA(SqlKind.EQUALS)) {
+          orSel += computeEqualsSelectivity(table, orPred, fieldNames);
+        } else if (orPred.isA(SqlKind.NOT_EQUALS)) {
+          orSel += 1.0 - computeEqualsSelectivity(table, orPred, fieldNames);
+        } else if (orPred.isA(SqlKind.LIKE)) {
+          // LIKE selectivity is 5% more than a similar equality predicate, capped at CALCITE guess
+          orSel +=  Math.min(computeEqualsSelectivity(table, orPred, fieldNames) + LIKE_PREDICATE_SELECTIVITY,
+              RelMdUtil.guessSelectivity(orPred));
+        } else if (orPred.isA(SqlKind.NOT)) {
           if (orPred instanceof RexCall) {
-            int colIdx = -1;
-            RexInputRef op = findRexInputRef(orPred);
-            if (op != null) {
-              colIdx = op.hashCode();
-            }
-            if (colIdx != -1 && colIdx < fieldNames.size()) {
-              String col = fieldNames.get(colIdx);
-              if (table.getStatsTable() != null
-                      && table.getStatsTable().getNdv(col) != null) {
-                orSel += 1.00 / table.getStatsTable().getNdv(col);
-              } else {
-                orSel += guess;
-              }
+            // LIKE selectivity is 5% more than a similar equality predicate, capped at CALCITE guess
+            RexNode childOp = ((RexCall) orPred).getOperands().get(0);
+            if (childOp.isA(SqlKind.LIKE)) {
+              orSel += 1.0 - Math.min(computeEqualsSelectivity(table, childOp, fieldNames) + LIKE_PREDICATE_SELECTIVITY,
+                      RelMdUtil.guessSelectivity(childOp));
             } else {
-              orSel += guess;
-              if (logger.isDebugEnabled()) {
-                logger.warn(String.format("No input reference $[%s] found for predicate [%s]",
-                    Integer.toString(colIdx), orPred.toString()));
-              }
+              orSel += 1.0 - RelMdUtil.guessSelectivity(orPred);
             }
           }
+        } else if (orPred.isA(SqlKind.IS_NULL)) {
+          orSel += computeNullSelectivity(table, orPred, fieldNames);
+        } else if (orPred.isA(SqlKind.IS_NOT_NULL)) {
+          orSel += 1.0 - computeNullSelectivity(table, orPred, fieldNames);
```

Comment:
If we already have a NN row count, the IS_NOT_NULL selectivity could be simplified to  NNRowCount/TotalRowCount.  Why do we need to do the (1 - NullSelectivity) ?   If we don't have the NN row count for any reason, then we are anyway defaulting to the Calcite guess.

Reply:
The NullSelectivity is essentially computed as 1-NNRowCount/TotalRowCount. So 1-NullSelectivity = NNRowCount/TotalRowCount; I was just reusing the function so we get `Cap` checks for free.

Reply:
Why not have a computeIsNotNullSelectivity() function ?  Sorry for the nit-pick but since we already have the exact NNRowCount to begin with, the clean formula without the subtraction would be preferred.

Reply:
Done